// @lay2/pw-core/build/module/models/raw-transaction.js
import PWCore from '../core';
import { SerializeRawTransaction } from '@ckb-lumos/types/lib/core';
import { validators, normalizers, Reader, transformers } from 'ckb-js-toolkit';
import { Blake2bHasher } from '../hashers';
export class RawTransaction {
    constructor(inputCells, outputs, cellDeps = [
        PWCore.config.defaultLock.cellDep,
        PWCore.config.pwLock.cellDep,
    ], headerDeps = [], version = '0x0') {
        this.inputCells = inputCells;
        this.outputs = outputs;
        this.cellDeps = cellDeps;
        this.headerDeps = headerDeps;
        this.version = version;
        this.inputs = inputCells.map((i) => i.toCellInput());
        this.outputsData = this.outputs.map((o) => o.getHexData());
    }
    sameWith(raw) {
        validators.ValidateTransaction(transformers.TransformTransaction(raw));
        return raw.toHash() === this.toHash();
    }
    toHash() {
        const rawTx = transformers.TransformRawTransaction(this);
        const hasher = new Blake2bHasher();
        return hasher
            .hash(new Reader(SerializeRawTransaction(normalizers.NormalizeRawTransaction(rawTx))))
            .serializeJson();
    }
    validate() {
        validators.ValidateRawTransaction(transformers.TransformRawTransaction(this));
        return true;
    }
    serializeJson() {
        return {
            version: this.version,
            cell_deps: this.cellDeps,
            header_deps: this.headerDeps,
            inputs: this.inputs,
            outputs: this.outputs,
            outputs_data: this.outputsData,
        };
    }
}
