// @lay2/pw-core/build/main/collectors/indexer-collector.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.IndexerCollector = void 0;
const sudt_collector_1 = require("./sudt-collector");
const __1 = require("..");
const ckb_indexer_1 = require("../helpers/ckb-indexer");
class IndexerCollector extends sudt_collector_1.SUDTCollector {
    constructor(apiBase) {
        super();
        this.apiBase = apiBase;
        this.indexer = new ckb_indexer_1.CkbIndexer(apiBase);
    }
    async getBalance(address) {
        const searchKey = {
            script: address.toLockScript().serializeJson(),
            script_type: ckb_indexer_1.ScriptType.lock,
            filter: {
                output_data_len_range: ['0x0', '0x1'],
            },
        };
        const cells = (await this.indexer.getCells(searchKey)).filter((cell) => cell.output.type === null);
        let balance = __1.Amount.ZERO;
        cells.forEach((cell) => {
            const amount = new __1.Amount(cell.output.capacity, __1.AmountUnit.shannon);
            balance = balance.add(amount);
        });
        return balance;
    }
    async collect(address, options) {
        if (!options || !options.neededAmount) {
            throw new Error("'neededAmount' in options must be provided");
        }
        let accCapacity = __1.Amount.ZERO;
        const terminator = (_index, cell) => {
            if (accCapacity.gte(options.neededAmount)) {
                return { stop: true, push: false };
            }
            if (cell.output_data.length / 2 - 1 > 0 || cell.output.type !== null) {
                return { stop: false, push: false };
            }
            else {
                accCapacity = accCapacity.add(new __1.Amount(cell.output.capacity, __1.AmountUnit.shannon));
                return { stop: false, push: true };
            }
        };
        const searchKey = {
            script: address.toLockScript().serializeJson(),
            script_type: ckb_indexer_1.ScriptType.lock,
            filter: {
                output_data_len_range: ['0x0', '0x1'],
            },
        };
        const cells = await this.indexer.getCells(searchKey, terminator);
        return cells.map((cell) => ckb_indexer_1.IndexerCellToCell(cell));
    }
    async getSUDTBalance(sudt, address) {
        const searchKey = {
            script: address.toLockScript().serializeJson(),
            script_type: ckb_indexer_1.ScriptType.lock,
            filter: {
                script: sudt.toTypeScript().serializeJson(),
            },
        };
        const cells = await this.indexer.getCells(searchKey);
        let balance = __1.Amount.ZERO;
        cells.forEach((cell) => {
            const amount = __1.Amount.fromUInt128LE(cell.output_data);
            balance = balance.add(amount);
        });
        return balance;
    }
    async collectSUDT(sudt, address, options) {
        if (!options || !options.neededAmount) {
            throw new Error("'neededAmount' in options must be provided");
        }
        const searchKey = {
            script: address.toLockScript().serializeJson(),
            script_type: ckb_indexer_1.ScriptType.lock,
            filter: {
                script: sudt.toTypeScript().serializeJson(),
            },
        };
        let accCapacity = __1.Amount.ZERO;
        const terminator = (_index, cell) => {
            if (accCapacity.gte(options.neededAmount)) {
                return { stop: true, push: false };
            }
            accCapacity = accCapacity.add(__1.Amount.fromUInt128LE(cell.output_data));
            return { stop: false, push: true };
        };
        const cells = await this.indexer.getCells(searchKey, terminator);
        return cells.map((cell) => ckb_indexer_1.IndexerCellToCell(cell));
    }
}
exports.IndexerCollector = IndexerCollector;
